### PR TITLE
Fix ua::NodeId::byte_string panic on NUL bytes

### LINKED ISSUE #338: valid `ByteString` node ids containing nul bytes induce a panic
The constructor for `ByteString` node ids via `ua::NodeId::byte_string` calls `CString::new`. `ByteString` node ids are allowed to contain nul bytes, but constructing a `ua::NodeId` from a buffer that contains a nul byte induces a panic.

### PR DESCRIPTION
**Pull request by cartercanedy**
Closes #338 
- Build byte-string node IDs directly from `ua::ByteString` instead of converting through a C string, so identifiers may contain embedded NUL bytes.
- Add regression coverage for NUL-containing byte-string node IDs, including parse/print round-tripping.

### CODE REVIEW

**Comment by cartercanedy** [comment]
@uklotzde I got a panic when attempting to integrate with a Wago PLC (PFC300, specifically) with a very simple object graph. Since opque ids aren't constrained by the spec at all, it's probably fair to say that there are other spec-compliant servers in the wild that would induce a panic, as well.

**Comment by uklotzde** [comment]
Thank you for contributing! Code looks reasonable at first sight. Having a regression test in place is helpful.

File: src/ua/data_types/node_id.rs [hunk 1]
```diff
@@ -77,34 +77,22 @@ impl NodeId {
...
-
-        // String allocation can fail but `UA_NODEID_BYTESTRING_ALLOC` does not tell us this when it
-        // happens. Instead, we end up with a well-defined node ID that has an empty byte string.
-        let inner = unsafe { UA_NODEID_BYTESTRING_ALLOC(ns_index, byte_string.as_ptr()) };
+        let byte_string = ua::ByteString::new(byte_string);
+        let mut node_id = Self::init();
+        node_id.0.namespaceIndex = ns_index;
+        node_id.0.identifierType = UA_NodeIdType::UA_NODEIDTYPE_BYTESTRING;
+        // SAFETY: We just selected the byte string identifier variant.
+        *unsafe { node_id.0.identifier.byteString.as_mut() } = byte_string.into_raw();
```
Comment (uklotzde) [inline src/ua/data_types/node_id.rs:90]: We should keep the debug assertion that `byte_string` is not null.

File: src/ua/data_types/node_id.rs [hunk 2]
```diff
@@ -340,4 +328,32 @@ mod tests {
         //
         let _node_id: ua::NodeId = "ns=0;i=2258".parse().expect("should be valid node ID");
     }
+
+    #[test]
+    fn byte_string_with_nul_bytes() {
```
Comment (uklotzde) [inline src/ua/data_types/node_id.rs:364]: While we are at it: How about an empty byte string?

**Comment by cartercanedy** [comment]
@uklotzde friendly ping :)

File: src/ua/data_types/node_id.rs [hunk 3]
```diff
@@ -77,34 +77,31 @@ impl NodeId {
...
-        let byte_string =
-            CString::new(byte_string).expect("node ID byte string does not contain NUL bytes");
-
-        // String allocation can fail but `UA_NODEID_BYTESTRING_ALLOC` does not tell us this when it
-        // happens. Instead, we end up with a well-defined node ID that has an empty byte string.
-        let inner = unsafe { UA_NODEID_BYTESTRING_ALLOC(ns_index, byte_string.as_ptr()) };
+    pub fn byte_string(ns_index: u16, bytes: &[u8]) -> Self {
+        let byte_string = ua::ByteString::new(bytes);
+
+        debug_assert!(
```
Comment (uklotzde) [inline src/ua/data_types/node_id.rs]: On second thought: This should already be guaranteed by `ua::ByteString::new()`. The debug assertion is actually not needed and might be confusing.

**Review by uklotzde — COMMENTED** [review]
Thanks for adding the extra test for the edge case.

Almost done. We should add a changelog entry for this bug fix.

File: src/ua/data_types/node_id.rs (hunk 3 above)
Comment (cartercanedy) [reply src/ua/data_types/node_id.rs]: My initial thought, as well, but I chose to keep as a sanity invariant check. Worth inlining up to the bytestring ctor, itself, though. I'll make that change

Comment (cartercanedy) [reply src/ua/data_types/node_id.rs]: ended up removing the assertion, full byte-wise cmp would be to imposing even in debug, imo

**Review by uklotzde — DISMISSED** [review]

File: CHANGELOG.md [hunk 4]
```diff
@@ -35,6 +35,7 @@ adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 - Fix linker errors for build target `x86_64-linux-unknown-gnu` by updating `open62541-sys` to
   version 0.5.4 ([#288](https://github.com/HMIProject/open62541/issues/288)).
 - Improve handling of and recovery from connection loss in `AsyncClient`.
+- Fix `ua::NodeId::byte_string` to allow byte strings that contain NUL bytes.
```
Comment (sgoll) [inline CHANGELOG.md]:
```suggestion
- Fix `ua::NodeId::byte_string()` to allow byte strings that contain NUL bytes.
```

File: src/ua/data_types/node_id.rs [hunk 5]
```diff
@@ -77,34 +77,25 @@ impl NodeId {
...
+
+        // SAFETY: We just selected the byte string identifier variant.
+        *unsafe { node_id.0.identifier.byteString.as_mut() } = byte_string.into_raw();
+
         debug_assert_eq!(
-            inner.identifierType,
+            node_id.0.identifierType,
             UA_NodeIdType::UA_NODEIDTYPE_BYTESTRING,
             "new node ID should have byte string type"
         );
```
Comment (sgoll) [inline src/ua/data_types/node_id.rs]: This is no longer necessary: Where the other methods use macros (`UA_NODEID_BYTESTRING_ALLOC`), we assign `identifierType` directly only three lines above. Checking what the macros do is reasonable, but confirming what we just did seems unnecessary 🙂

```suggestion
```

**Review by sgoll — CHANGES_REQUESTED** [review]

File: src/ua/data_types/node_id.rs (hunk 5 above)
Comment (uklotzde) [reply src/ua/data_types/node_id.rs]: @cartercanedy Please delete the debug assertion manually. This suggestion contains a deleted line that cannot be applied.

**Review by sgoll — APPROVED** [review]
Thank you for your contribution. Everything is looking good now and I am going to merge this pull request 🚀
